src/blog/oilpan-pointer-compression.md

```diff
@@ -0,0 +1,247 @@
+
+Having this usage in mind, we were looking for a scheme that transparently handles case 1.-3. Since the compression and decompression sequence will be inlined everywhere Member is used, the following properties are also desirable:
+
+- Fast and compact instruction sequence to minimize icache misses.
+- Branchless instruction sequence to avoid using up branch predictors.
+
+Since it is expected that reads significantly outnumber writes, we allow for an asymmetric scheme where fast decompression is preferred.
+
+### Compression and decompression
+
+For brevity, this description only covers the final compression scheme used. See our [design doc](https://docs.google.com/document/d/1neGN8Jq-1JLrWK3cvwRIfrSjLgE0Srjv-uq7Ze38Iao) for more information on how we got there and the alternatives considered.
+
+The main idea for the scheme that is implemented as of today is to separate regular heap pointers from `nullptr` and sentinel by relying on alignment of the heap cage.  Essentially, the heap cage is allocated with alignment such that the least significant bit of the upper halfword is always set.  We denote the upper and lower half (32 bits each) as U<sub>31</sub>...U<sub>0</sub> and L<sub>31</sub>...L<sub>0</sub>, respectively.
+
+:::table-wrapper
+<!-- markdownlint-disable no-inline-html -->
+|              | upper half                               | lower half                                |
+| ------------ | ---------------------------------------: | ----------------------------------------: |
+| heap pointer | <tt>U<sub>31</sub>...U<sub>1</sub>1</tt> | <tt>L<sub>31</sub>...L<sub>2</sub>00</tt> |
+| `nullptr`    | <tt>0...0</tt>                           | <tt>0...000</tt>                          |
+| sentinel     | <tt>0...0</tt>                           | <tt>0...010</tt>                          |
+<!-- markdownlint-enable no-inline-html -->
+:::
+
+Compression generates a compressed value by merely right-shifting by one and truncating away the upper half of the value.  In this way, the alignment bit (which now becomes the most significant bit of the compressed value) signals a valid heap pointer.
+
+:::table-wrapper
+| C++                                             | x64 assembly  |
+| :---------------------------------------------- | :------------ |
+| ```cpp                                          | ```asm        \
+| uint32_t Compress(void* ptr) {                  | mov rax, rdi  \
+|   return ((uintptr_t)ptr) >> 1;                 | shr rax       \
+| }                                               | ```           \
+| ```                                             |               |
+:::
+
+The encoding for compressed values is thus as follows:
+
+:::table-wrapper
+<!-- markdownlint-disable no-inline-html -->
+|              | compressed value                          |
+| ------------ | ----------------------------------------: |
+| heap pointer | <tt>1L<sub>31</sub>...L<sub>2</sub>0</tt> |
```

Comment:
Please change
"<tt>1L<sub>31</sub>...L<sub>2</sub>0</tt>"
to
"<tt>1L<sub>31</sub>...L<sub>2</sub>00</tt>"